src/main/kotlin/org/malachite/estella/services/ApplicationService.kt

```diff
@@ -15,43 +14,42 @@ import java.util.*
 @Service
 class ApplicationService(
     @Autowired private val applicationRepository: ApplicationRepository,
-    @Autowired private val offerRepository: OfferRepository,
-    @Autowired private val jobSeekerRepository: JobSeekerRepository,
+    @Autowired private val offerService: OfferService,
+    @Autowired private val jobSeekerService: JobSeekerService,
     @Autowired private val recruitmentProcessService: RecruitmentProcessService,
     @Autowired private val interviewService: InterviewService,
     @Autowired private val mailService: MailService
-) {
-    fun insertApplicationLoggedInUser(offerId: Int, jobSeeker: JobSeeker, applicationPayload: ApplicationLoggedInPayload): Application {
-        val offer = offerRepository.findById(offerId).get()
+) : EStellaService<Application>() {
+
+    override val throwable: Exception = ApplicationNotFoundException()
+
+    fun insertApplicationLoggedInUser(
+        offerId: Int,
+        jobSeeker: JobSeeker,
+        applicationPayload: ApplicationLoggedInPayload
+    ): Application {
+        val offer = offerService.getOffer(offerId)
         val stage = offer.recruitmentProcess?.stages?.getOrNull(0)
         return stage?.let {
-            val application = applicationRepository.save(
-                applicationPayload.toApplication(it, jobSeeker)
-            )
-            if (application.seekerFiles.isNotEmpty()) {
-                val updatedSeekerFiles = jobSeeker.files.plus(application.seekerFiles)
-                val updatedJobSeeker = jobSeeker.copy(files = updatedSeekerFiles)
-                jobSeekerRepository.save(updatedJobSeeker)
-            }
+            val application = applicationRepository.save(applicationPayload.toApplication(it, jobSeeker))
+            if (application.seekerFiles.isNotEmpty())
+                jobSeeker.files.plus(application.seekerFiles)
+                    .let { jobSeekerService.updateJobSeekerFiles(jobSeeker, it) }
             mailService.sendApplicationConfirmationMail(offer, application)
             application
-        } ?: throw NoSuchElementException()
+        } ?: throw throwable
```

Comment:
My tu chyba nie powinniśmy rzucać throwable, a `StageNotFoundException` (nawet jeśli czegoś takiego nie ma, to powinno być) - ogólnie ten elvis operator odnosi się do tego, że stage jest nullowy, nie aplikacja (co w sumie by było dziwne, jak insertujesz aplikację, a tu Ci wali, że nie ma takiej aplikacji - no tak, ma rację, bo ją właśnie chcę dodać)

Reply:
Zastanawiałem się nad tym i przy obecnym podejściu że każda proces rekrutacyjny musi mieć przynajmniej dwa stage (Applied i Ended) to taki wyjątek jest kompletnie bez sensu, bo bliżej prawdy byłoby wysłanie internal server error XDD